tests/helpers/test_config_entry_oauth2_flow.py

```diff
@@ -420,6 +420,94 @@ async def test_oauth_session(hass, flow_handler, local_impl, aioclient_mock):
     assert round(config_entry.data["token"]["expires_at"] - now) == 100
 
 
+async def test_oauth_session_with_clock_slightly_out_of_sync(
+    hass, flow_handler, local_impl, aioclient_mock
+):
+    """Test the OAuth2 session helper when the remote clock is slightly out of sync."""
+    flow_handler.async_register_implementation(hass, local_impl)
+
+    aioclient_mock.post(
+        TOKEN_URL, json={"access_token": ACCESS_TOKEN_2, "expires_in": 19}
+    )
+
+    aioclient_mock.post("https://example.com", status=201)
+
+    config_entry = MockConfigEntry(
+        domain=TEST_DOMAIN,
+        data={
+            "auth_implementation": TEST_DOMAIN,
+            "token": {
+                "refresh_token": REFRESH_TOKEN,
+                "access_token": ACCESS_TOKEN_1,
+                "expires_in": 19,
+                "expires_at": time.time() + 19,  # Forces a refresh,
+                "token_type": "bearer",
+                "random_other_data": "should_stay",
+            },
+        },
+    )
+
+    now = time.time()
+    session = config_entry_oauth2_flow.OAuth2Session(hass, config_entry, local_impl)
+    resp = await session.async_request("post", "https://example.com")
+    assert resp.status == 201
+
+    # Refresh token, make request
+    assert len(aioclient_mock.mock_calls) == 2
+
+    assert (
+        aioclient_mock.mock_calls[1][3]["authorization"] == f"Bearer {ACCESS_TOKEN_2}"
+    )
+
+    assert config_entry.data["token"]["refresh_token"] == REFRESH_TOKEN
+    assert config_entry.data["token"]["access_token"] == ACCESS_TOKEN_2
+    assert config_entry.data["token"]["expires_in"] == 19
+    assert config_entry.data["token"]["random_other_data"] == "should_stay"
+    assert round(config_entry.data["token"]["expires_at"] - now) == 19
```

Comment:
Will this be racy/flaky? Not sure if there is a convention in home assistant for handling this type of thing already, or if its typical to just assume the code above runs fast enough.

Reply:
Well above code is non racy. Since there is no threads involved here and no awaits.

I suspect the refresh token code may be racy and probably needs some async lock to be safe. But would be for another pull.

Reply:
Good ooint.  I was assuming the two time.time() calls could theoretically happy far enough apart and reusing `now` across both would be safer, but it looks like a tiny gap and I'm probably being too paranoid here.